File: togo/internal/format/javadoc/writer.go
// Port of src/format/javadoc/writer.ts (gjf's JavadocWriter).
//
// Stateful renderer: accepts requests/writes and produces wrapped javadoc.
// Classic `/** */` only; Markdown methods kept for fidelity but unexercised.

package javadoc

import "strings"

type wsKind int

const (
	wsNone wsKind = iota
	wsWhitespace
	wsNewline
	wsBlankLine
)

var backslashLiteral = Token{kind: literal, value: "\\"}

type javadocWriter struct {
	blockIndent int

	out                                 strings.Builder
	continuingListItemOfInnermostList   bool
	continuingFooterTag                 bool
	continuingListItemStack             intNestingStack
	continuingListStack                 intNestingStack
	postWriteModifiedContinuingListStck intNestingStack
	remainingOnLine                     int
	atStartOfLine                       bool
	requestedWhitespace                 wsKind
	wroteAnythingSignificant            bool
}

func newWriter(blockIndent int) *javadocWriter {
	return &javadocWriter{blockIndent: blockIndent}
}

func (w *javadocWriter) requestWhitespace() { w.request(wsWhitespace) }

func (w *javadocWriter) request(ws wsKind) {
	if ws > w.requestedWhitespace {
		w.requestedWhitespace = ws
	}
}

func (w *javadocWriter) requestBlankLine() { w.request(wsBlankLine) }
func (w *javadocWriter) requestNewline()   { w.request(wsNewline) }

func (w *javadocWriter) writeBeginJavadoc() {
	w.out.WriteString("/**")
	w.writeNewline(true)
}

func (w *javadocWriter) writeEndJavadoc() {
	w.out.WriteString("\n")
	w.appendSpaces(w.blockIndent + 1)
	w.out.WriteString("*/")
}

func (w *javadocWriter) writeFooterJavadocTagStart(t Token) {
	w.continuingListItemOfInnermostList = false
	w.continuingListItemStack.reset()
	w.continuingListStack.reset()
	w.postWriteModifiedContinuingListStck.reset()
	switch {
	case !w.wroteAnythingSignificant:
		// Javadoc consists solely of tags (OK for @Override).
	case !w.continuingFooterTag:
		w.requestBlankLine()
	default:
		w.continuingFooterTag = false
		w.requestNewline()
	}
	w.writeToken(t)
	w.continuingFooterTag = true
}

func (w *javadocWriter) writeSnippetBegin(t Token) {
	w.requestBlankLine()
	w.writeToken(t)
}

func (w *javadocWriter) writeSnippetEnd(t Token) {
	w.writeToken(t)
	w.requestBlankLine()
}

func (w *javadocWriter) writeListOpen(t Token) {
	w.requestBlankLine()
	w.writeToken(t)
	w.continuingListItemOfInnermostList = false
	indent := 0
	if t.value != "" {
		indent = 2
	}
	w.continuingListStack.push(indent)
	w.postWriteModifiedContinuingListStck.push(1)
	w.requestNewline()
}

func (w *javadocWriter) writeListClose(t Token) {
	w.requestNewline()
	w.continuingListItemStack.popIfNotEmpty()
	w.continuingListStack.popIfNotEmpty()
	w.writeToken(t)
	w.postWriteModifiedContinuingListStck.popIfNotEmpty()
	w.requestBlankLine()
}

func (w *javadocWriter) writeListItemOpen(t Token) {
	w.requestNewline()
	if w.continuingListItemOfInnermostList {
		w.continuingListItemOfInnermostList = false
		w.continuingListItemStack.popIfNotEmpty()
	}
	w.writeToken(t)
	w.continuingListItemOfInnermostList = true
	w.continuingListItemStack.push(len(t.value))
}

func (w *javadocWriter) writeHeaderOpen(t Token) {
	if w.wroteAnythingSignificant {
		w.requestBlankLine()
	}
	w.writeToken(t)
}

func (w *javadocWriter) writeHeaderClose(t Token) {
	w.writeToken(t)
	w.requestBlankLine()
}

func (w *javadocWriter) writeParagraphOpen(t Token) {
	if !w.wroteAnythingSignificant {
		return
	}
	w.requestBlankLine()
	w.writeToken(t)
}

func (w *javadocWriter) writeBlockquoteOpenOrClose(t Token) {
	w.requestBlankLine()
	w.writeToken(t)
	w.requestBlankLine()
}

func (w *javadocWriter) writePreOpen(t Token) {
	w.requestBlankLine()
	w.writeToken(t)
}

func (w *javadocWriter) writePreClose(t Token) {
	w.writeToken(t)
	w.requestBlankLine()
}

func (w *javadocWriter) writeCodeOpen(t Token)  { w.writeToken(t) }
func (w *javadocWriter) writeCodeClose(t Token) { w.writeToken(t) }

func (w *javadocWriter) writeTableOpen(t Token) {
	w.requestBlankLine()
	w.writeToken(t)
}

func (w *javadocWriter) writeTableClose(t Token) {
	w.writeToken(t)
	w.requestBlankLine()
}

func (w *javadocWriter) writeHTMLComment(t Token) {
	w.requestNewline()
	w.writeToken(t)
	w.requestNewline()
}

func (w *javadocWriter) writeBr(t Token) {
	w.writeToken(t)
	w.requestNewline()
}

func (w *javadocWriter) writeMoeEndStripComment(t Token) {
	w.writeNewline(false)
	w.writeToken(t)
	w.requestNewline()
}

func (w *javadocWriter) requestMoeBeginStripComment(_ Token) { w.requestNewline() }

func (w *javadocWriter) writeLineBreakNoAutoIndent() { w.writeNewline(false) }

func (w *javadocWriter) writeMarkdownHardLineBreak() {
	w.writeLiteral(backslashLiteral)
	w.writeNewline(true)
}

func (w *javadocWriter) writeLiteral(t Token) { w.writeToken(t) }

func (w *javadocWriter) writeMarkdownFencedCodeBlock(t Token) {
	w.flushWhitespace()
	w.out.WriteString(t.value)
	w.requestBlankLine()
}

func (w *javadocWriter) writeMarkdownTable(t Token) {
	w.flushWhitespace()
	lines := strings.Split(t.value, "\n")
	w.out.WriteString(lines[0])
	for _, line := range lines[1:] {
		w.writeNewline(false)
		w.out.WriteString(line)
	}
	w.requestBlankLine()
}

func (w *javadocWriter) String() string { return w.out.String() }

func (w *javadocWriter) flushWhitespace() {
	if w.requestedWhitespace == wsBlankLine &&
		(!w.postWriteModifiedContinuingListStck.isEmpty() || w.continuingFooterTag) {
		w.requestedWhitespace = wsNewline
	}
	switch w.requestedWhitespace {
	case wsBlankLine:
		w.writeBlankLine()
		w.requestedWhitespace = wsNone
	case wsNewline:
		w.writeNewline(true)
		w.requestedWhitespace = wsNone
	}
}

func (w *javadocWriter) writeToken(t Token) {
	if t.value == "" {
		return
	}
	w.flushWhitespace()
	needWhitespace := w.requestedWhitespace == wsWhitespace

	extra := 0
	if needWhitespace {
		extra = 1
	}
	if !w.atStartOfLine && len(t.value)+extra > w.remainingOnLine {
		w.writeNewline(true)
	}
	if !w.atStartOfLine && needWhitespace {
		w.out.WriteString(" ")
		w.remainingOnLine--
	}

	w.out.WriteString(t.value)
	if !isStartOfLine(t.kind) {
		w.atStartOfLine = false
	}
	w.remainingOnLine -= len(t.value)
	w.requestedWhitespace = wsNone
	w.wroteAnythingSignificant = true
}

func (w *javadocWriter) writeNewlineStart() {
	w.out.WriteString("\n")
	w.appendSpaces(w.blockIndent + 1)
	w.out.WriteString("*")
}

func (w *javadocWriter) writeBlankLine() {
	w.writeNewlineStart()
	w.writeNewline(true)
}

func (w *javadocWriter) writeNewline(autoIndent bool) {
	w.writeNewlineStart()
	w.appendSpaces(1)
	w.remainingOnLine = maxLineLength - w.blockIndent - 3
	if autoIndent {
		w.appendSpaces(w.innerIndent())
		w.remainingOnLine -= w.innerIndent()
	}
	w.atStartOfLine = true
}

func (w *javadocWriter) innerIndent() int {
	n := w.continuingListItemStack.total + w.continuingListStack.total
	if w.continuingFooterTag {
		n += 4
	}
	return n
}

func (w *javadocWriter) appendSpaces(count int) { w.out.WriteString(strings.Repeat(" ", count)) }

File: src/format/javadoc/writer.ts
// Port of google-java-format core/.../java/javadoc/JavadocWriter.java.
//
// Stateful renderer: accepts "requests" and "writes" from the formatter and
// produces wrapped javadoc. It tracks list/footer nesting and remaining columns
// so it can re-wrap prose. We only support classic `/** */` javadoc (the Markdown
// `///` methods are kept for fidelity but are not exercised yet).

import { MAX_LINE_LENGTH } from "./formatter.ts";
import { IntNestingStack } from "./nesting-stack.ts";
import { isStartOfLine, type Token } from "./token.ts";

const enum WS {
  None,
  Whitespace,
  Newline,
  BlankLine,
}

const BACKSLASH_LITERAL: Token = { kind: "literal", value: "\\" };

export class JavadocWriter {
  private out = "";
  private continuingListItemOfInnermostList = false;
  private continuingFooterTag = false;
  private readonly continuingListItemStack = new IntNestingStack();
  private readonly continuingListStack = new IntNestingStack();
  private readonly postWriteModifiedContinuingListStack = new IntNestingStack();
  private remainingOnLine = 0;
  private atStartOfLine = false;
  private requestedWhitespace = WS.None;
  private wroteAnythingSignificant = false;

  constructor(
    private readonly blockIndent: number,
    private readonly classicJavadoc = true,
  ) {}

  requestWhitespace(): void {
    this.request(WS.Whitespace);
  }

  private request(ws: WS): void {
    if (ws > this.requestedWhitespace) this.requestedWhitespace = ws;
  }

  private requestBlankLine(): void {
    this.request(WS.BlankLine);
  }

  private requestNewline(): void {
    this.request(WS.Newline);
  }

  writeBeginJavadoc(): void {
    this.out += "/**";
    this.writeNewline();
  }

  writeEndJavadoc(): void {
    this.out += "\n";
    this.appendSpaces(this.blockIndent + 1);
    this.out += "*/";
  }

  writeFooterJavadocTagStart(token: Token): void {
    this.continuingListItemOfInnermostList = false;
    this.continuingListItemStack.reset();
    this.continuingListStack.reset();
    this.postWriteModifiedContinuingListStack.reset();
    if (!this.wroteAnythingSignificant) {
      // Javadoc consists solely of tags (OK for @Override).
    } else if (!this.continuingFooterTag) {
      this.requestBlankLine();
    } else {
      this.continuingFooterTag = false;
      this.requestNewline();
    }
    this.writeToken(token);
    this.continuingFooterTag = true;
  }

  writeSnippetBegin(token: Token): void {
    this.requestBlankLine();
    this.writeToken(token);
  }

  writeSnippetEnd(token: Token): void {
    this.writeToken(token);
    this.requestBlankLine();
  }

  writeListOpen(token: Token): void {
    this.requestBlankLine();
    this.writeToken(token);
    this.continuingListItemOfInnermostList = false;
    this.continuingListStack.push(token.value === "" ? 0 : 2);
    this.postWriteModifiedContinuingListStack.push();
    this.requestNewline();
  }

  writeListClose(token: Token): void {
    this.requestNewline();
    this.continuingListItemStack.popIfNotEmpty();
    this.continuingListStack.popIfNotEmpty();
    this.writeToken(token);
    this.postWriteModifiedContinuingListStack.popIfNotEmpty();
    this.requestBlankLine();
  }

  writeListItemOpen(token: Token): void {
    this.requestNewline();
    if (this.continuingListItemOfInnermostList) {
      this.continuingListItemOfInnermostList = false;
      this.continuingListItemStack.popIfNotEmpty();
    }
    this.writeToken(token);
    this.continuingListItemOfInnermostList = true;
    this.continuingListItemStack.push(token.value.length);
  }

  writeHeaderOpen(token: Token): void {
    if (this.wroteAnythingSignificant) this.requestBlankLine();
    this.writeToken(token);
  }

  writeHeaderClose(token: Token): void {
    this.writeToken(token);
    this.requestBlankLine();
  }

  writeParagraphOpen(token: Token): void {
    if (!this.wroteAnythingSignificant) return; // ignore a leading <p>
    this.requestBlankLine();
    this.writeToken(token);
  }

  writeBlockquoteOpenOrClose(token: Token): void {
    this.requestBlankLine();
    this.writeToken(token);
    this.requestBlankLine();
  }

  writePreOpen(token: Token): void {
    this.requestBlankLine();
    this.writeToken(token);
  }

  writePreClose(token: Token): void {
    this.writeToken(token);
    this.requestBlankLine();
  }

  writeCodeOpen(token: Token): void {
    this.writeToken(token);
  }

  writeCodeClose(token: Token): void {
    this.writeToken(token);
  }

  writeTableOpen(token: Token): void {
    this.requestBlankLine();
    this.writeToken(token);
  }

  writeTableClose(token: Token): void {
    this.writeToken(token);
    this.requestBlankLine();
  }

  writeHtmlComment(token: Token): void {
    this.requestNewline();
    this.writeToken(token);
    this.requestNewline();
  }

  writeBr(token: Token): void {
    this.writeToken(token);
    this.requestNewline();
  }

  writeMoeEndStripComment(token: Token): void {
    this.writeLineBreakNoAutoIndent();
    this.writeToken(token);
    this.requestNewline();
  }

  requestMoeBeginStripComment(_token: Token): void {
    // MOE strip comments are not used; request handled minimally.
    this.requestNewline();
  }

  writeLineBreakNoAutoIndent(): void {
    this.writeNewline(false);
  }

  writeMarkdownHardLineBreak(): void {
    this.writeLiteral(BACKSLASH_LITERAL);
    this.writeNewline();
  }

  writeLiteral(token: Token): void {
    this.writeToken(token);
  }

  // Markdown-only; inert for classic javadoc but kept for fidelity.
  writeMarkdownFencedCodeBlock(token: Token): void {
    this.flushWhitespace();
    this.out += token.value;
    this.requestBlankLine();
  }

  writeMarkdownTable(token: Token): void {
    this.flushWhitespace();
    const lines = token.value.split("\n");
    this.out += lines[0];
    for (const line of lines.slice(1)) {
      this.writeNewline(false);
      this.out += line;
    }
    this.requestBlankLine();
  }

  toString(): string {
    return this.out;
  }

  private flushWhitespace(): void {
    if (
      this.classicJavadoc &&
      this.requestedWhitespace === WS.BlankLine &&
      (!this.postWriteModifiedContinuingListStack.isEmpty() || this.continuingFooterTag)
    ) {
      // No blank lines inside lists or footer tags.
      this.requestedWhitespace = WS.Newline;
    }
    if (this.requestedWhitespace === WS.BlankLine) {
      this.writeBlankLine();
      this.requestedWhitespace = WS.None;
    } else if (this.requestedWhitespace === WS.Newline) {
      this.writeNewline();
      this.requestedWhitespace = WS.None;
    }
  }

  private writeToken(token: Token): void {
    if (token.value === "") return;
    this.flushWhitespace();
    const needWhitespace = this.requestedWhitespace === WS.Whitespace;

    if (
      !this.atStartOfLine &&
      token.value.length + (needWhitespace ? 1 : 0) > this.remainingOnLine
    ) {
      this.writeNewline();
    }
    if (!this.atStartOfLine && needWhitespace) {
      this.out += " ";
      this.remainingOnLine--;
    }

    this.out += token.value;
    if (!isStartOfLine(token.kind)) this.atStartOfLine = false;
    this.remainingOnLine -= token.value.length;
    this.requestedWhitespace = WS.None;
    this.wroteAnythingSignificant = true;
  }

  private writeNewlineStart(): void {
    this.out += "\n";
    this.appendSpaces(this.blockIndent + 1);
    this.out += "*";
  }

  private writeBlankLine(): void {
    this.writeNewlineStart();
    this.writeNewline();
  }

  private writeNewline(autoIndent = true): void {
    this.writeNewlineStart();
    this.appendSpaces(1);
    this.remainingOnLine = MAX_LINE_LENGTH - this.blockIndent - 3;
    if (autoIndent) {
      this.appendSpaces(this.innerIndent());
      this.remainingOnLine -= this.innerIndent();
    }
    this.atStartOfLine = true;
  }

  private innerIndent(): number {
    let n = this.continuingListItemStack.total() + this.continuingListStack.total();
    if (this.continuingFooterTag) n += 4;
    return n;
  }

  private appendSpaces(count: number): void {
    this.out += " ".repeat(count);
  }
}
